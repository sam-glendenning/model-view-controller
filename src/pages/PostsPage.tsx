import React from 'react';
import {
  Alert,
  Box,
  CircularProgress,
  Container,
  Fab,
  Snackbar,
  Typography,
} from '@mui/material';
import { Add } from '@mui/icons-material';
import Post from '@/features/posts';
import { useGetPostsController } from '@/features/posts/view/useGetPostsController';
import { CreatePostDialog } from '@/features/posts/create/CreatePostDialog';
import { useCreatePostController } from '@/features/posts/create/useCreatePostController';

export const PostsPage: React.FC = () => {
  const {
    posts,
    postsLoading,
    postsError,
    snackbar,
    closeSnackbar,
    showErrorMessage,
    onPostDeleted,
    onPostUpdated,
    onPostCreated,
  } = useGetPostsController();

  const {
    postData,
    isCreatePostDialogOpen,
    hideCreatePostDialog,
    updateTitle,
    updateBody,
    updateUserId,
    confirmCreate,
    isCreating,
    showCreatePostDialog,
  } = useCreatePostController({
    onPostCreated,
    onPostCreationError: (_error: Error) =>
      showErrorMessage('Failed to create post'),
  });

  if (postsLoading) {
    return (
      <Container
        maxWidth="md"
        sx={{ py: 4, display: 'flex', justifyContent: 'center' }}
      >
        <CircularProgress />
      </Container>
    );
  }

  if (postsError) {
    return (
      <Container maxWidth="md" sx={{ py: 4 }}>
        <Alert severity="error">
          Error loading posts: {postsError.message}
        </Alert>
      </Container>
    );
  }

  return (
    <Container maxWidth="lg" sx={{ py: 4 }}>
      <Box sx={{ mb: 4 }}>
        <Typography variant="h3" component="h1" gutterBottom>
          Posts Management
        </Typography>
        <Typography variant="h6" color="text.secondary">
          Demonstrating MVC architecture with React, TanStack Query, and
          Material-UI
        </Typography>
      </Box>

      {posts && posts.length > 0 && (
        <Box>
          {posts.map(post => (
            <Post
              key={post.id}
              postData={post}
              onPostDeleted={onPostDeleted}
              onPostUpdated={onPostUpdated}
            />
          ))}
        </Box>
      )}

      <Fab color="primary" aria-label="add post" onClick={showCreatePostDialog}>
        <Add />
      </Fab>

      <CreatePostDialog
        open={isCreatePostDialogOpen}
        onClose={hideCreatePostDialog}
        postData={postData}
        onTitleChange={updateTitle}
        onBodyChange={updateBody}
        onUserIdChange={updateUserId}
        onSubmit={confirmCreate}
        isCreating={isCreating}
      />

      <Snackbar
        open={snackbar.open}
        autoHideDuration={4000}
        onClose={closeSnackbar}
      >
        <Alert
          onClose={closeSnackbar}
          severity={snackbar.severity}
          sx={{ width: '100%' }}
        >
          {snackbar.message}
        </Alert>
      </Snackbar>
    </Container>
  );
};
