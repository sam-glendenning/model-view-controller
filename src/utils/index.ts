export * from './validation';
export * from './errorHandling';
