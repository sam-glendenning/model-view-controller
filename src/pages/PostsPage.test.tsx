import { screen, waitFor, within, act } from '@testing-library/react';
import userEvent, { type UserEvent } from '@testing-library/user-event';
import { render } from '@/test/utils';
import { PostsPage } from './PostsPage';
import { server } from '@/test/mocks/server';
import { http, HttpResponse } from 'msw';

describe('PostsPage Integration Tests', () => {
  let user: UserEvent;

  beforeEach(() => {
    user = userEvent.setup();
  });

  const createView = () => render(<PostsPage />);

  it('renders correctly', async () => {
    const { container } = createView();

    await waitFor(() => {
      expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });

  it('should show loading state initially', () => {
    createView();

    expect(screen.getByRole('progressbar')).toBeInTheDocument();
  });

  it('should show error state when API fails', async () => {
    // Mock API failure
    server.use(
      http.get('https://jsonplaceholder.typicode.com/posts', () => {
        return new HttpResponse(null, { status: 500 });
      }),
    );

    createView();

    await waitFor(() => {
      expect(screen.getByText(/Error loading posts/)).toBeInTheDocument();
    });

    // Reset handlers
    server.resetHandlers();
  });

  it('should create new post', async () => {
    createView();

    // Wait for posts to load
    await waitFor(() => {
      expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
    });

    // Open create dialog
    const openCreatePostDialogButton = screen.getByRole('button', {
      name: 'add post',
    });
    await user.click(openCreatePostDialogButton);

    await waitFor(() => {
      expect(screen.getByRole('dialog')).toBeInTheDocument();
    });
    const dialog = screen.getByRole('dialog');

    // Fill form
    const titleInput = within(dialog).getByRole('textbox', { name: 'Title' });
    const bodyInput = within(dialog).getByRole('textbox', { name: 'Content' });

    await user.type(titleInput, 'New Test Post');
    await user.type(bodyInput, 'New test post body');

    // Submit form
    const createButton = within(dialog).getByRole('button', { name: 'Create' });

    await act(async () => await user.click(createButton));

    // Wait for success message
    await waitFor(() => {
      expect(
        screen.getByText(/Post \d+ created successfully!/),
      ).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });

    // Note: In tests with MSW, the new post won't actually appear in the list
    // because MSW handlers use isolated mock data. The important part is that
    // the create flow works and shows the success message.
  });

  it('should edit existing post', async () => {
    createView();

    // Wait for posts to load
    await waitFor(() => {
      expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
    });

    // Find and click edit button for first post
    const editButtons = screen.getAllByRole('button', { name: 'edit post' });
    await user.click(editButtons[0]);

    // Edit the title
    const titleInput = screen.getByRole('textbox', { name: 'Title' });
    await user.clear(titleInput);
    await user.type(titleInput, 'Updated Test Post Title');

    // Save changes
    const saveButton = screen.getByRole('button', { name: 'Save' });
    await act(async () => await user.click(saveButton));

    // Wait for success message
    await waitFor(() => {
      expect(
        screen.getByText(/Post 1 updated successfully!/),
      ).toBeInTheDocument();
    });
  });

  it('should delete existing post', async () => {
    createView();

    // Wait for posts to load
    await waitFor(() => {
      expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
    });

    // Find and click delete button for first post
    const deleteButtons = screen.getAllByRole('button', {
      name: 'delete post',
    });
    await user.click(deleteButtons[0]);

    // Wait for delete confirmation dialog
    await waitFor(() => {
      expect(screen.getByRole('dialog')).toBeInTheDocument();
    });

    const dialog = screen.getByRole('dialog');
    expect(within(dialog).getByText('Delete Post')).toBeInTheDocument();
    const confirmDeleteButton = within(dialog).getByRole('button', {
      name: 'Confirm',
    });

    // Confirm deletion
    await act(async () => await user.click(confirmDeleteButton));

    // Wait for success message
    await waitFor(() => {
      expect(
        screen.getByText(/Post 1 deleted successfully!/),
      ).toBeInTheDocument();
    });
  });

  it('should close snackbar when close button is clicked', async () => {
    createView();

    // Wait for posts to load
    await waitFor(() => {
      expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
    });

    // Trigger an action to show snackbar (delete post)
    const deleteButtons = screen.getAllByRole('button', {
      name: 'delete post',
    });
    await user.click(deleteButtons[0]);

    // Wait for delete confirmation dialog
    await waitFor(() => {
      expect(screen.getByRole('dialog')).toBeInTheDocument();
    });

    const dialog = screen.getByRole('dialog');
    expect(within(dialog).getByText('Delete Post')).toBeInTheDocument();
    const confirmDeleteButton = within(dialog).getByRole('button', {
      name: 'Confirm',
    });
    await act(async () => await user.click(confirmDeleteButton));

    // Wait for success message
    await waitFor(() => {
      expect(
        screen.getByText(/Post 1 deleted successfully!/),
      ).toBeInTheDocument();
    });

    // Close snackbar
    const closeButton = screen.getByRole('button', { name: 'Close' });
    await user.click(closeButton);

    // Wait for snackbar to disappear
    await waitFor(() => {
      expect(
        screen.queryByText(/Post 1 deleted successfully!/),
      ).not.toBeInTheDocument();
    });
  });

  it('should handle create post error', async () => {
    // Mock API failure for create
    server.use(
      http.post('https://jsonplaceholder.typicode.com/posts', () => {
        return new HttpResponse(null, { status: 500 });
      }),
    );

    createView();

    // Wait for posts to load
    await waitFor(() => {
      expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
    });

    // Open create dialog
    const openCreatePostDialogButton = screen.getByRole('button', {
      name: 'add post',
    });
    await user.click(openCreatePostDialogButton);

    await waitFor(() => {
      expect(screen.getByRole('dialog')).toBeInTheDocument();
    });
    const dialog = screen.getByRole('dialog');

    // Fill form
    const titleInput = within(dialog).getByRole('textbox', { name: 'Title' });
    const bodyInput = within(dialog).getByRole('textbox', { name: 'Content' });

    await user.type(titleInput, 'New Test Post');
    await user.type(bodyInput, 'New test post body');

    // Submit form
    const createButton = within(dialog).getByRole('button', { name: 'Create' });
    await act(async () => await user.click(createButton));

    // Wait for error message
    await waitFor(() => {
      expect(screen.getByText('Failed to create post')).toBeInTheDocument();
    });

    // Reset handlers
    server.resetHandlers();
  });
});
