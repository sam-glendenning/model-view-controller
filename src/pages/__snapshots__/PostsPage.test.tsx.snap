// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsPage Integration Tests renders correctly 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-1sqq6so-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-1yjvs5a"
    >
      <h1
        class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom css-1mktp8m-MuiTypography-root"
      >
        Posts Management
      </h1>
      <h6
        class="MuiTypography-root MuiTypography-h6 css-1jgmn5n-MuiTypography-root"
      >
        Demonstrating MVC architecture with React, TanStack Query, and Material-UI
      </h6>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiCard-root css-okv3ew-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiBox-root css-0"
            >
              <div
                class="MuiBox-root css-21y1cj"
              >
                <h3
                  class="MuiTypography-root MuiTypography-h6 css-yxjt1e-MuiTypography-root"
                >
                  Test Post Title
                </h3>
                <div
                  class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorPrimary MuiChip-outlinedPrimary css-1y6wsfa-MuiChip-root"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall css-rrn746-MuiChip-label"
                  >
                    Post #1
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root MuiTypography-body2 css-eb0ott-MuiTypography-root"
              >
                This is a test post body with some content to display.
              </p>
              <div
                class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorSecondary MuiChip-outlinedSecondary css-1sr0oy8-MuiChip-root"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall css-rrn746-MuiChip-label"
                >
                  User 1
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1j9t6id-MuiCardActions-root"
          >
            <div
              class="MuiBox-root css-1i27l4i"
            >
              <button
                aria-label="edit post"
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary css-1k23hlb-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="EditIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                    />
                  </svg>
                </span>
                Edit
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <button
              aria-label="delete post"
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorError MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorError css-1ir7th6-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                  />
                </svg>
              </span>
              Delete
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiCard-root css-okv3ew-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiBox-root css-0"
            >
              <div
                class="MuiBox-root css-21y1cj"
              >
                <h3
                  class="MuiTypography-root MuiTypography-h6 css-yxjt1e-MuiTypography-root"
                >
                  Another Test Post
                </h3>
                <div
                  class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorPrimary MuiChip-outlinedPrimary css-1y6wsfa-MuiChip-root"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall css-rrn746-MuiChip-label"
                  >
                    Post #2
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root MuiTypography-body2 css-eb0ott-MuiTypography-root"
              >
                This is another test post with different content.
              </p>
              <div
                class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorSecondary MuiChip-outlinedSecondary css-1sr0oy8-MuiChip-root"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall css-rrn746-MuiChip-label"
                >
                  User 1
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1j9t6id-MuiCardActions-root"
          >
            <div
              class="MuiBox-root css-1i27l4i"
            >
              <button
                aria-label="edit post"
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary css-1k23hlb-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="EditIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                    />
                  </svg>
                </span>
                Edit
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <button
              aria-label="delete post"
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorError MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorError css-1ir7th6-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                  />
                </svg>
              </span>
              Delete
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiCard-root css-okv3ew-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiBox-root css-0"
            >
              <div
                class="MuiBox-root css-21y1cj"
              >
                <h3
                  class="MuiTypography-root MuiTypography-h6 css-yxjt1e-MuiTypography-root"
                >
                  Jane's Post
                </h3>
                <div
                  class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorPrimary MuiChip-outlinedPrimary css-1y6wsfa-MuiChip-root"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall css-rrn746-MuiChip-label"
                  >
                    Post #3
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root MuiTypography-body2 css-eb0ott-MuiTypography-root"
              >
                A post written by Jane Smith about various topics.
              </p>
              <div
                class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorSecondary MuiChip-outlinedSecondary css-1sr0oy8-MuiChip-root"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall css-rrn746-MuiChip-label"
                >
                  User 2
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1j9t6id-MuiCardActions-root"
          >
            <div
              class="MuiBox-root css-1i27l4i"
            >
              <button
                aria-label="edit post"
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary css-1k23hlb-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="EditIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                    />
                  </svg>
                </span>
                Edit
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <button
              aria-label="delete post"
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorError MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorError css-1ir7th6-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                  />
                </svg>
              </span>
              Delete
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiCard-root css-okv3ew-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiBox-root css-0"
            >
              <div
                class="MuiBox-root css-21y1cj"
              >
                <h3
                  class="MuiTypography-root MuiTypography-h6 css-yxjt1e-MuiTypography-root"
                >
                  Understanding React Hooks
                </h3>
                <div
                  class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorPrimary MuiChip-outlinedPrimary css-1y6wsfa-MuiChip-root"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall css-rrn746-MuiChip-label"
                  >
                    Post #4
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root MuiTypography-body2 css-eb0ott-MuiTypography-root"
              >
                React Hooks have revolutionized how we write components. This post explores the most commonly used hooks and their practical applications.
              </p>
              <div
                class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorSecondary MuiChip-outlinedSecondary css-1sr0oy8-MuiChip-root"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall css-rrn746-MuiChip-label"
                >
                  User 1
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1j9t6id-MuiCardActions-root"
          >
            <div
              class="MuiBox-root css-1i27l4i"
            >
              <button
                aria-label="edit post"
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary css-1k23hlb-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="EditIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                    />
                  </svg>
                </span>
                Edit
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <button
              aria-label="delete post"
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorError MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorError css-1ir7th6-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                  />
                </svg>
              </span>
              Delete
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiCard-root css-okv3ew-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiBox-root css-0"
            >
              <div
                class="MuiBox-root css-21y1cj"
              >
                <h3
                  class="MuiTypography-root MuiTypography-h6 css-yxjt1e-MuiTypography-root"
                >
                  Building Scalable APIs
                </h3>
                <div
                  class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorPrimary MuiChip-outlinedPrimary css-1y6wsfa-MuiChip-root"
                >
                  <span
                    class="MuiChip-label MuiChip-labelSmall css-rrn746-MuiChip-label"
                  >
                    Post #5
                  </span>
                </div>
              </div>
              <p
                class="MuiTypography-root MuiTypography-body2 css-eb0ott-MuiTypography-root"
              >
                A comprehensive guide to building APIs that can handle growth and maintain performance as your application scales.
              </p>
              <div
                class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorSecondary MuiChip-outlinedSecondary css-1sr0oy8-MuiChip-root"
              >
                <span
                  class="MuiChip-label MuiChip-labelSmall css-rrn746-MuiChip-label"
                >
                  User 2
                </span>
              </div>
            </div>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-1j9t6id-MuiCardActions-root"
          >
            <div
              class="MuiBox-root css-1i27l4i"
            >
              <button
                aria-label="edit post"
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorPrimary css-1k23hlb-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="EditIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
                    />
                  </svg>
                </span>
                Edit
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <button
              aria-label="delete post"
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorError MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-colorError css-1ir7th6-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                  />
                </svg>
              </span>
              Delete
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <button
      aria-label="add post"
      class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary css-1fo2sjy-MuiButtonBase-root-MuiFab-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
